## 1.IMPORTING PANDAS


In [51]:
import pandas as pd

In [16]:
import requests
from io import StringIO

url = "https://raw.githubusercontent.com/cs109/2014_data/master/countries.csv"
headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:66.0) Gecko/20100101 Firefox/66.0"}
req = requests.get(url, headers=headers)
data = StringIO(req.text)

pd.read_csv(data)

Country         Region
0      Algeria         AFRICA
1       Angola         AFRICA
2        Benin         AFRICA
3     Botswana         AFRICA
4      Burkina         AFRICA
..         ...            ...
189   Paraguay  SOUTH AMERICA
190       Peru  SOUTH AMERICA
191   Suriname  SOUTH AMERICA
192    Uruguay  SOUTH AMERICA
193  Venezuela  SOUTH AMERICA

[194 rows x 2 columns]

# 4.sep parameter

##### generally in read_csv the sep parameter = ',' but when we use like tab seperated values file then to differentiate between column  we use this sep parameter as shown below

In [24]:
pd.read_csv('movie_titles_metadata.tsv')

m0\t10 things i hate about you\t1999\t6.90\t62847\t['comedy' 'romance']
0    m1\t1492: conquest of paradise\t1992\t6.20\t10...                     
1    m2\t15 minutes\t2001\t6.10\t25854\t['action' '...                     
2    m3\t2001: a space odyssey\t1968\t8.40\t163227\...                     
3    m4\t48 hrs.\t1982\t6.90\t22289\t['action' 'com...                     
4    m5\tthe fifth element\t1997\t7.50\t133756\t['a...                     
..                                                 ...                     
611  m612\twatchmen\t2009\t7.80\t135229\t['action' ...                     
612  m613\txxx\t2002\t5.60\t53505\t['action' 'adven...                     
613  m614\tx-men\t2000\t7.40\t122149\t['action' 'sc...                     
614  m615\tyoung frankenstein\t1974\t8.00\t57618\t[...                     
615  m616\tzulu dawn\t1979\t6.40\t1911\t['action' '...                     

[616 rows x 1 columns]

##### we are not able to understand what is written above b/c all columns have been mixed up b/c there were no comma in the file from where read_csv can separate on basis of comma so we will use sep parameter as here the values are separated by tab 

In [29]:
pd.read_csv('movie_titles_metadata.tsv',sep='\t')

m0  10 things i hate about you  1999  6.90     62847  \
0      m1  1492: conquest of paradise  1992   6.2   10421.0   
1      m2                  15 minutes  2001   6.1   25854.0   
2      m3       2001: a space odyssey  1968   8.4  163227.0   
3      m4                     48 hrs.  1982   6.9   22289.0   
4      m5           the fifth element  1997   7.5  133756.0   
..    ...                         ...   ...   ...       ...   
611  m612                    watchmen  2009   7.8  135229.0   
612  m613                         xxx  2002   5.6   53505.0   
613  m614                       x-men  2000   7.4  122149.0   
614  m615          young frankenstein  1974   8.0   57618.0   
615  m616                   zulu dawn  1979   6.4    1911.0   

                                  ['comedy' 'romance']  
0          ['adventure' 'biography' 'drama' 'history']  
1                ['action' 'crime' 'drama' 'thriller']  
2                     ['adventure' 'mystery' 'sci-fi']  
3       ['action' 'comedy' 'crime' 'drama' 'thriller']  
4    ['action' 'adventure' 'romance' 'sci-fi' 'thri...  
..                                                 ...  
611  ['action' 'crime' 'fantasy' 'mystery' 'sci-fi'...  
612                     ['action' 'adventure' 'crime']  
613                                ['action' 'sci-fi']  
614                                ['comedy' 'sci-fi']  
615     ['action' 'adventure' 'drama' 'history' 'war']  

[616 rows x 6 columns]

#### in the above o/p we can see that one of the rows only have been named as column name which is wrong so we will use names parameter to correct it there in names parameter we will pass a list of columns

In [34]:
pd.read_csv('movie_titles_metadata.tsv',sep='\t',names=['sr_no','name','release_year','ratings','votes','genres'])

sr_no                        name release_year  ratings     votes  \
0      m0  10 things i hate about you         1999      6.9   62847.0   
1      m1  1492: conquest of paradise         1992      6.2   10421.0   
2      m2                  15 minutes         2001      6.1   25854.0   
3      m3       2001: a space odyssey         1968      8.4  163227.0   
4      m4                     48 hrs.         1982      6.9   22289.0   
..    ...                         ...          ...      ...       ...   
612  m612                    watchmen         2009      7.8  135229.0   
613  m613                         xxx         2002      5.6   53505.0   
614  m614                       x-men         2000      7.4  122149.0   
615  m615          young frankenstein         1974      8.0   57618.0   
616  m616                   zulu dawn         1979      6.4    1911.0   

                                                genres  
0                                 ['comedy' 'romance']  
1          ['adventure' 'biography' 'drama' 'history']  
2                ['action' 'crime' 'drama' 'thriller']  
3                     ['adventure' 'mystery' 'sci-fi']  
4       ['action' 'comedy' 'crime' 'drama' 'thriller']  
..                                                 ...  
612  ['action' 'crime' 'fantasy' 'mystery' 'sci-fi'...  
613                     ['action' 'adventure' 'crime']  
614                                ['action' 'sci-fi']  
615                                ['comedy' 'sci-fi']  
616     ['action' 'adventure' 'drama' 'history' 'war']  

[617 rows x 6 columns]

# 5.index_col parameter

##### if there are useless columns and we dont want to use it like pandas create automatic indexing but we dont want it so we can remove it 

In [38]:
pd.read_csv('movie_titles_metadata.tsv',sep='\t',names=['sr_no','name','release_year','ratings','votes','genres'],index_col='sr_no')

name release_year  ratings     votes  \
sr_no                                                               
m0     10 things i hate about you         1999      6.9   62847.0   
m1     1492: conquest of paradise         1992      6.2   10421.0   
m2                     15 minutes         2001      6.1   25854.0   
m3          2001: a space odyssey         1968      8.4  163227.0   
m4                        48 hrs.         1982      6.9   22289.0   
...                           ...          ...      ...       ...   
m612                     watchmen         2009      7.8  135229.0   
m613                          xxx         2002      5.6   53505.0   
m614                        x-men         2000      7.4  122149.0   
m615           young frankenstein         1974      8.0   57618.0   
m616                    zulu dawn         1979      6.4    1911.0   

                                                  genres  
sr_no                                                     
m0                                  ['comedy' 'romance']  
m1           ['adventure' 'biography' 'drama' 'history']  
m2                 ['action' 'crime' 'drama' 'thriller']  
m3                      ['adventure' 'mystery' 'sci-fi']  
m4        ['action' 'comedy' 'crime' 'drama' 'thriller']  
...                                                  ...  
m612   ['action' 'crime' 'fantasy' 'mystery' 'sci-fi'...  
m613                      ['action' 'adventure' 'crime']  
m614                                 ['action' 'sci-fi']  
m615                                 ['comedy' 'sci-fi']  
m616      ['action' 'adventure' 'drama' 'history' 'war']  

[617 rows x 5 columns]

# 6. Header Parameter

In [41]:
pd.read_csv('test.csv')

Unnamed: 0   Unnamed: 1 Unnamed: 2              Unnamed: 3 Unnamed: 4  \
0           0  enrollee_id       city  city_development_index     gender   
1           1        29725    city_40                   0.776       Male   
2           2        11561    city_21                   0.624        NaN   
3           3        33241   city_115                   0.789        NaN   
4           4          666   city_162                   0.767       Male   

                Unnamed: 5           Unnamed: 6       Unnamed: 7  \
0      relevent_experience  enrolled_university  education_level   
1   No relevent experience        no_enrollment         Graduate   
2   No relevent experience     Full time course         Graduate   
3   No relevent experience                  NaN         Graduate   
4  Has relevent experience        no_enrollment          Masters   

         Unnamed: 8  Unnamed: 9   Unnamed: 10     Unnamed: 11   Unnamed: 12  \
0  major_discipline  experience  company_size    company_type  last_new_job   
1              STEM          15         50-99         Pvt Ltd            >4   
2              STEM           5           NaN             NaN         never   
3   Business Degree          <1           NaN         Pvt Ltd         never   
4              STEM         >20         50-99  Funded Startup             4   

      Unnamed: 13 Unnamed: 14  
0  training_hours      target  
1              47           0  
2              83           0  
3              52           1  
4               8           0

### in the above examples pandas donot able to understand which one is header so we will going to define it by using parameter haeder=1 1 b/c mere dataset mein 1st row pe hi wo column nmae tha

In [46]:
pd.read_csv('test.csv', header = 1)

0  enrollee_id      city  city_development_index gender  \
0  1        29725   city_40                   0.776   Male   
1  2        11561   city_21                   0.624    NaN   
2  3        33241  city_115                   0.789    NaN   
3  4          666  city_162                   0.767   Male   

       relevent_experience enrolled_university education_level  \
0   No relevent experience       no_enrollment        Graduate   
1   No relevent experience    Full time course        Graduate   
2   No relevent experience                 NaN        Graduate   
3  Has relevent experience       no_enrollment         Masters   

  major_discipline experience company_size    company_type last_new_job  \
0             STEM         15        50-99         Pvt Ltd           >4   
1             STEM          5          NaN             NaN        never   
2  Business Degree         <1          NaN         Pvt Ltd        never   
3             STEM        >20        50-99  Funded Startup            4   

   training_hours  target  
0              47       0  
1              83       0  
2              52       1  
3               8       0

# 7. usecols parameter

#### if we want specific columns only at the time of import or after import so we can specify the column name and get that specific columns data only which we require

In [55]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

In [63]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

##### if we want only columns of enrollee_id , 	gender,education_level,training_hours,target then we will use the parameter usecols=['column_names,,'c_n','c_n']) 

In [60]:
pd.read_csv('aug_train.csv',usecols=['gender','education_level','training_hours','target'] )

gender education_level  training_hours  target
0       Male        Graduate              36     1.0
1       Male        Graduate              47     0.0
2        NaN        Graduate              83     0.0
3        NaN        Graduate              52     1.0
4       Male         Masters               8     0.0
...      ...             ...             ...     ...
19153   Male        Graduate              42     1.0
19154   Male        Graduate              52     1.0
19155   Male        Graduate              44     0.0
19156   Male     High School              97     0.0
19157    NaN  Primary School             127     0.0

[19158 rows x 4 columns]

# 8. Skiprows/nrows parameters

##### if we not want to take some rows then we will use skiprows and if we want to take some particular or specific rows then we will use nrows

In [67]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

In [71]:
pd.read_csv('aug_train.csv',skiprows=[1,5])

enrollee_id      city  city_development_index gender  \
0            29725   city_40                   0.776   Male   
1            11561   city_21                   0.624    NaN   
2            33241  city_115                   0.789    NaN   
3            21651  city_176                   0.764    NaN   
4            28806  city_160                   0.920   Male   
...            ...       ...                     ...    ...   
19151         7386  city_173                   0.878   Male   
19152        31398  city_103                   0.920   Male   
19153        24576  city_103                   0.920   Male   
19154         5756   city_65                   0.802   Male   
19155        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0       No relevent experience       no_enrollment        Graduate   
1       No relevent experience    Full time course        Graduate   
2       No relevent experience                 NaN        Graduate   
3      Has relevent experience    Part time course        Graduate   
4      Has relevent experience       no_enrollment     High School   
...                        ...                 ...             ...   
19151   No relevent experience       no_enrollment        Graduate   
19152  Has relevent experience       no_enrollment        Graduate   
19153  Has relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment     High School   
19155   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM         15        50-99         Pvt Ltd           >4   
1                 STEM          5          NaN             NaN        never   
2      Business Degree         <1          NaN         Pvt Ltd        never   
3                 STEM         11          NaN             NaN            1   
4                  NaN          5        50-99  Funded Startup            1   
...                ...        ...          ...             ...          ...   
19151       Humanities         14          NaN             NaN            1   
19152             STEM         14          NaN             NaN            4   
19153             STEM        >20        50-99         Pvt Ltd            4   
19154              NaN         <1      500-999         Pvt Ltd            2   
19155              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  47     0.0  
1                  83     0.0  
2                  52     1.0  
3                  24     1.0  
4                  24     0.0  
...               ...     ...  
19151              42     1.0  
19152              52     1.0  
19153              44     0.0  
19154              97     0.0  
19155             127     0.0  

[19156 rows x 14 columns]

In [73]:
pd.read_csv('aug_train.csv',nrows=1000)

enrollee_id      city  city_development_index  gender  \
0           8949  city_103                   0.920    Male   
1          29725   city_40                   0.776    Male   
2          11561   city_21                   0.624     NaN   
3          33241  city_115                   0.789     NaN   
4            666  city_162                   0.767    Male   
..           ...       ...                     ...     ...   
995        28500   city_21                   0.624     NaN   
996        10371  city_103                   0.920  Female   
997        10028   city_73                   0.754    Male   
998        29671   city_40                   0.776    Male   
999         4482  city_103                   0.920    Male   

         relevent_experience enrolled_university education_level  \
0    Has relevent experience       no_enrollment        Graduate   
1     No relevent experience       no_enrollment        Graduate   
2     No relevent experience    Full time course        Graduate   
3     No relevent experience                 NaN        Graduate   
4    Has relevent experience       no_enrollment         Masters   
..                       ...                 ...             ...   
995  Has relevent experience    Full time course        Graduate   
996   No relevent experience       no_enrollment             Phd   
997  Has relevent experience       no_enrollment        Graduate   
998   No relevent experience    Full time course        Graduate   
999  Has relevent experience       no_enrollment        Graduate   

    major_discipline experience company_size    company_type last_new_job  \
0               STEM        >20          NaN             NaN            1   
1               STEM         15        50-99         Pvt Ltd           >4   
2               STEM          5          NaN             NaN        never   
3    Business Degree         <1          NaN         Pvt Ltd        never   
4               STEM        >20        50-99  Funded Startup            4   
..               ...        ...          ...             ...          ...   
995             STEM         <1      500-999         Pvt Ltd            1   
996             STEM          7      500-999   Public Sector            4   
997             STEM         15    1000-4999         Pvt Ltd            1   
998             STEM          9          NaN             NaN            1   
999             STEM        >20        10/49  Funded Startup            1   

     training_hours  target  
0                36     1.0  
1                47     0.0  
2                83     0.0  
3                52     1.0  
4                 8     0.0  
..              ...     ...  
995              22     0.0  
996              28     0.0  
997              42     0.0  
998              50     0.0  
999             105     0.0  

[1000 rows x 14 columns]

# 9. encoding parameter

##### generally UTF-8 encoding is present but if there are different encoding than the utf-8 then it can cause problem to our csv reading therfore we have to specify the encoding type there only


In [77]:
pd.read_csv('zomato.csv')

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 7044: invalid continuation byte

###### here encoding was in latin-1 so we declare while loading the dataset

In [79]:
pd.read_csv('zomato.csv',encoding='latin-1')

Restaurant ID           Restaurant Name  Country Code              City  \
0           6317637          Le Petit Souffle           162       Makati City   
1           6304287          Izakaya Kikufuji           162       Makati City   
2           6300002    Heat - Edsa Shangri-La           162  Mandaluyong City   
3           6318506                      Ooma           162  Mandaluyong City   
4           6314302               Sambo Kojin           162  Mandaluyong City   
...             ...                       ...           ...               ...   
9546        5915730              NamlÛ± Gurme           208         ÛÁstanbul   
9547        5908749             Ceviz AÛôacÛ±           208         ÛÁstanbul   
9548        5915807                     Huqqa           208         ÛÁstanbul   
9549        5916112              Aôôk Kahve           208         ÛÁstanbul   
9550        5927402  Walter's Coffee Roastery           208         ÛÁstanbul   

                                                Address  \
0     Third Floor, Century City Mall, Kalayaan Avenu...   
1     Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2     Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3     Third Floor, Mega Fashion Hall, SM Megamall, O...   
4     Third Floor, Mega Atrium, SM Megamall, Ortigas...   
...                                                 ...   
9546  Kemankeô Karamustafa Paôa Mahallesi, RÛ±htÛ±...   
9547  Koôuyolu Mahallesi, Muhittin íìstí_ndaÛô Cadd...   
9548  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9549  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9550  CafeaÛôa Mahallesi, BademaltÛ± Sokak, No 21/B,...   

                                        Locality  \
0      Century City Mall, Poblacion, Makati City   
1     Little Tokyo, Legaspi Village, Makati City   
2     Edsa Shangri-La, Ortigas, Mandaluyong City   
3         SM Megamall, Ortigas, Mandaluyong City   
4         SM Megamall, Ortigas, Mandaluyong City   
...                                          ...   
9546                                    Karakí_y   
9547                                   Koôuyolu   
9548                                 Kuruí_eôme   
9549                                 Kuruí_eôme   
9550                                        Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall, Poblacion, Makati City, Mak...  121.027535   
1     Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101   
2     Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831   
3     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475   
4     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508   
...                                                 ...         ...   
9546                                Karakí_y, ÛÁstanbul   28.977392   
9547                               Koôuyolu, ÛÁstanbul   29.041297   
9548                             Kuruí_eôme, ÛÁstanbul   29.034640   
9549                             Kuruí_eôme, ÛÁstanbul   29.036019   
9550                                    Moda, ÛÁstanbul   29.026016   

       Latitude                          Cuisines  ...          Currency  \
0     14.565443        French, Japanese, Desserts  ...  Botswana Pula(P)   
1     14.553708                          Japanese  ...  Botswana Pula(P)   
2     14.581404  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)   
3     14.585318                   Japanese, Sushi  ...  Botswana Pula(P)   
4     14.584450                  Japanese, Korean  ...  Botswana Pula(P)   
...         ...                               ...  ...               ...   
9546  41.022793                           Turkish  ...  Turkish Lira(TL)   
9547  41.009847   World Cuisine, Patisserie, Cafe  ...  Turkish Lira(TL)   
9548  41.055817            Italian, World Cuisine  ...  Turkish Lira(TL)   
9549  41.057979                   Restaurant Cafe  ...  Turkish Lira(TL)   
9550  40.9

# 10. skip bad lines parameter

### let say i have 100 rows X 14 columns of dataset in which in the row number 14 there are 20 columns which is wrong so we will skipp that row using skip bad line parameter as this can cause problem in our algorithm

### we will use like pd.read_csv('filepath',error_bad_lines=False) it will delete all the rows which contains extra columns

# 11. dtype parameters

In [90]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

In [92]:
pd.read_csv('aug_train.csv').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19158 entries, 0 to 19157
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   enrollee_id             19158 non-null  int64  
 1   city                    19158 non-null  object 
 2   city_development_index  19158 non-null  float64
 3   gender                  14650 non-null  object 
 4   relevent_experience     19158 non-null  object 
 5   enrolled_university     18772 non-null  object 
 6   education_level         18698 non-null  object 
 7   major_discipline        16345 non-null  object 
 8   experience              19093 non-null  object 
 9   company_size            13220 non-null  object 
 10  company_type            13018 non-null  object 
 11  last_new_job            18735 non-null  object 
 12  training_hours          19158 non-null  int64  
 13  target                  19158 non-null  float64
dtypes: float64(2), int64(2), object(10)
me

##### here we can see the data type of target is float which can be done to integer as in float it is using extra memory 

In [97]:
pd.read_csv('aug_train.csv',dtype={'target':int})

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36       1  
1                  47       0  
2                  83       0  
3                  52       1  
4                   8       0  
...               ...     ...  
19153              42       1  
19154              52       1  
19155              44       0  
19156              97       0  
19157             127       0  

[19158 rows x 14 columns]

In [103]:
pd.read_csv('aug_train.csv',dtype={'target':int}).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19158 entries, 0 to 19157
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   enrollee_id             19158 non-null  int64  
 1   city                    19158 non-null  object 
 2   city_development_index  19158 non-null  float64
 3   gender                  14650 non-null  object 
 4   relevent_experience     19158 non-null  object 
 5   enrolled_university     18772 non-null  object 
 6   education_level         18698 non-null  object 
 7   major_discipline        16345 non-null  object 
 8   experience              19093 non-null  object 
 9   company_size            13220 non-null  object 
 10  company_type            13018 non-null  object 
 11  last_new_job            18735 non-null  object 
 12  training_hours          19158 non-null  int64  
 13  target                  19158 non-null  int32  
dtypes: float64(1), int32(1), int64(2), obj

# 12. Handling dates

## like if we have date columns which is of datatype object(string) so we cannot perform various operations of date on that particular columns therefore we will use parse_dates=['column_names']

In [108]:
pd.read_csv('IPL Matches 2008-2020.csv')

id        city        date player_of_match  \
0     335982   Bangalore  2008-04-18     BB McCullum   
1     335983  Chandigarh  2008-04-19      MEK Hussey   
2     335984       Delhi  2008-04-19     MF Maharoof   
3     335985      Mumbai  2008-04-20      MV Boucher   
4     335986     Kolkata  2008-04-20       DJ Hussey   
..       ...         ...         ...             ...   
811  1216547       Dubai  2020-09-28  AB de Villiers   
812  1237177       Dubai  2020-11-05       JJ Bumrah   
813  1237178   Abu Dhabi  2020-11-06   KS Williamson   
814  1237180   Abu Dhabi  2020-11-08      MP Stoinis   
815  1237181       Dubai  2020-11-10        TA Boult   

                                          venue  neutral_venue  \
0                         M Chinnaswamy Stadium              0   
1    Punjab Cricket Association Stadium, Mohali              0   
2                              Feroz Shah Kotla              0   
3                              Wankhede Stadium              0   
4                                  Eden Gardens              0   
..                                          ...            ...   
811         Dubai International Cricket Stadium              0   
812         Dubai International Cricket Stadium              0   
813                        Sheikh Zayed Stadium              0   
814                        Sheikh Zayed Stadium              0   
815         Dubai International Cricket Stadium              0   

                           team1                        team2  \
0    Royal Challengers Bangalore        Kolkata Knight Riders   
1                Kings XI Punjab          Chennai Super Kings   
2               Delhi Daredevils             Rajasthan Royals   
3                 Mumbai Indians  Royal Challengers Bangalore   
4          Kolkata Knight Riders              Deccan Chargers   
..                           ...                          ...   
811  Royal Challengers Bangalore               Mumbai Indians   
812               Mumbai Indians               Delhi Capitals   
813  Royal Challengers Bangalore          Sunrisers Hyderabad   
814               Delhi Capitals          Sunrisers Hyderabad   
815               Delhi Capitals               Mumbai Indians   

                     toss_winner toss_decision                       winner  \
0    Royal Challengers Bangalore         field        Kolkata Knight Riders   
1            Chennai Super Kings           bat          Chennai Super Kings   
2               Rajasthan Royals           bat             Delhi Daredevils   
3                 Mumbai Indians           bat  Royal Challengers Bangalore   
4                Deccan Chargers           bat        Kolkata Knight Riders   
..                           ...           ...                          ...   
811               Mumbai Indians         field  Royal Challengers Bangalore   
812               Delhi Capitals         field               Mumbai Indians   
813          Sunrisers Hyderabad         field          Sunrisers Hyderabad   
814               Delhi Capitals           bat               Delhi Capitals   
815               Delhi Capitals           bat               Mumbai Indians   

      result  result_margin eliminator method      umpire1         umpire2  
0       runs          140.0          N    NaN    Asad Rauf     RE Koertzen  
1       runs           33.0          N    NaN    MR Benson      SL Shastri  
2    wickets            9.0          N    NaN    Aleem Dar  GA Pratapkumar  
3    wickets            5.0          N    NaN     SJ Davis       DJ Harper  
4    wickets            5.0          N    NaN    BF Bowden     K Hariharan  
..       ...            ...        ...    ...          ...             ...  
811      tie            NaN          Y    NaN  Nitin Menon      PR Reiffel  
812     runs           57.0          N    NaN  CB Gaffaney     Nitin Menon  
813  wickets            6.0          N    NaN   PR Reiffel          S Ravi  
814     runs           17.0          N    Na

In [110]:
pd.read_csv('IPL Matches 2008-2020.csv').info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               816 non-null    int64  
 1   city             803 non-null    object 
 2   date             816 non-null    object 
 3   player_of_match  812 non-null    object 
 4   venue            816 non-null    object 
 5   neutral_venue    816 non-null    int64  
 6   team1            816 non-null    object 
 7   team2            816 non-null    object 
 8   toss_winner      816 non-null    object 
 9   toss_decision    816 non-null    object 
 10  winner           812 non-null    object 
 11  result           812 non-null    object 
 12  result_margin    799 non-null    float64
 13  eliminator       812 non-null    object 
 14  method           19 non-null     object 
 15  umpire1          816 non-null    object 
 16  umpire2          816 non-null    object 
dtypes: float64(1), i

In [112]:
pd.read_csv('IPL Matches 2008-2020.csv',parse_dates=['date'])

id        city       date player_of_match  \
0     335982   Bangalore 2008-04-18     BB McCullum   
1     335983  Chandigarh 2008-04-19      MEK Hussey   
2     335984       Delhi 2008-04-19     MF Maharoof   
3     335985      Mumbai 2008-04-20      MV Boucher   
4     335986     Kolkata 2008-04-20       DJ Hussey   
..       ...         ...        ...             ...   
811  1216547       Dubai 2020-09-28  AB de Villiers   
812  1237177       Dubai 2020-11-05       JJ Bumrah   
813  1237178   Abu Dhabi 2020-11-06   KS Williamson   
814  1237180   Abu Dhabi 2020-11-08      MP Stoinis   
815  1237181       Dubai 2020-11-10        TA Boult   

                                          venue  neutral_venue  \
0                         M Chinnaswamy Stadium              0   
1    Punjab Cricket Association Stadium, Mohali              0   
2                              Feroz Shah Kotla              0   
3                              Wankhede Stadium              0   
4                                  Eden Gardens              0   
..                                          ...            ...   
811         Dubai International Cricket Stadium              0   
812         Dubai International Cricket Stadium              0   
813                        Sheikh Zayed Stadium              0   
814                        Sheikh Zayed Stadium              0   
815         Dubai International Cricket Stadium              0   

                           team1                        team2  \
0    Royal Challengers Bangalore        Kolkata Knight Riders   
1                Kings XI Punjab          Chennai Super Kings   
2               Delhi Daredevils             Rajasthan Royals   
3                 Mumbai Indians  Royal Challengers Bangalore   
4          Kolkata Knight Riders              Deccan Chargers   
..                           ...                          ...   
811  Royal Challengers Bangalore               Mumbai Indians   
812               Mumbai Indians               Delhi Capitals   
813  Royal Challengers Bangalore          Sunrisers Hyderabad   
814               Delhi Capitals          Sunrisers Hyderabad   
815               Delhi Capitals               Mumbai Indians   

                     toss_winner toss_decision                       winner  \
0    Royal Challengers Bangalore         field        Kolkata Knight Riders   
1            Chennai Super Kings           bat          Chennai Super Kings   
2               Rajasthan Royals           bat             Delhi Daredevils   
3                 Mumbai Indians           bat  Royal Challengers Bangalore   
4                Deccan Chargers           bat        Kolkata Knight Riders   
..                           ...           ...                          ...   
811               Mumbai Indians         field  Royal Challengers Bangalore   
812               Delhi Capitals         field               Mumbai Indians   
813          Sunrisers Hyderabad         field          Sunrisers Hyderabad   
814               Delhi Capitals           bat               Delhi Capitals   
815               Delhi Capitals           bat               Mumbai Indians   

      result  result_margin eliminator method      umpire1         umpire2  
0       runs          140.0          N    NaN    Asad Rauf     RE Koertzen  
1       runs           33.0          N    NaN    MR Benson      SL Shastri  
2    wickets            9.0          N    NaN    Aleem Dar  GA Pratapkumar  
3    wickets            5.0          N    NaN     SJ Davis       DJ Harper  
4    wickets            5.0          N    NaN    BF Bowden     K Hariharan  
..       ...            ...        ...    ...          ...             ...  
811      tie            NaN          Y    NaN  Nitin Menon      PR Reiffel  
812     runs           57.0          N    NaN  CB Gaffaney     Nitin Menon  
813  wickets            6.0          N    NaN   PR Reiffel          S Ravi  
814     runs           17.0          N    NaN   PR Reiff

In [114]:
pd.read_csv('IPL Matches 2008-2020.csv',parse_dates=['date']).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   id               816 non-null    int64         
 1   city             803 non-null    object        
 2   date             816 non-null    datetime64[ns]
 3   player_of_match  812 non-null    object        
 4   venue            816 non-null    object        
 5   neutral_venue    816 non-null    int64         
 6   team1            816 non-null    object        
 7   team2            816 non-null    object        
 8   toss_winner      816 non-null    object        
 9   toss_decision    816 non-null    object        
 10  winner           812 non-null    object        
 11  result           812 non-null    object        
 12  result_margin    799 non-null    float64       
 13  eliminator       812 non-null    object        
 14  method           19 non-null     object   

# 13. converters

In [117]:
pd.read_csv('IPL Matches 2008-2020.csv')

id        city        date player_of_match  \
0     335982   Bangalore  2008-04-18     BB McCullum   
1     335983  Chandigarh  2008-04-19      MEK Hussey   
2     335984       Delhi  2008-04-19     MF Maharoof   
3     335985      Mumbai  2008-04-20      MV Boucher   
4     335986     Kolkata  2008-04-20       DJ Hussey   
..       ...         ...         ...             ...   
811  1216547       Dubai  2020-09-28  AB de Villiers   
812  1237177       Dubai  2020-11-05       JJ Bumrah   
813  1237178   Abu Dhabi  2020-11-06   KS Williamson   
814  1237180   Abu Dhabi  2020-11-08      MP Stoinis   
815  1237181       Dubai  2020-11-10        TA Boult   

                                          venue  neutral_venue  \
0                         M Chinnaswamy Stadium              0   
1    Punjab Cricket Association Stadium, Mohali              0   
2                              Feroz Shah Kotla              0   
3                              Wankhede Stadium              0   
4                                  Eden Gardens              0   
..                                          ...            ...   
811         Dubai International Cricket Stadium              0   
812         Dubai International Cricket Stadium              0   
813                        Sheikh Zayed Stadium              0   
814                        Sheikh Zayed Stadium              0   
815         Dubai International Cricket Stadium              0   

                           team1                        team2  \
0    Royal Challengers Bangalore        Kolkata Knight Riders   
1                Kings XI Punjab          Chennai Super Kings   
2               Delhi Daredevils             Rajasthan Royals   
3                 Mumbai Indians  Royal Challengers Bangalore   
4          Kolkata Knight Riders              Deccan Chargers   
..                           ...                          ...   
811  Royal Challengers Bangalore               Mumbai Indians   
812               Mumbai Indians               Delhi Capitals   
813  Royal Challengers Bangalore          Sunrisers Hyderabad   
814               Delhi Capitals          Sunrisers Hyderabad   
815               Delhi Capitals               Mumbai Indians   

                     toss_winner toss_decision                       winner  \
0    Royal Challengers Bangalore         field        Kolkata Knight Riders   
1            Chennai Super Kings           bat          Chennai Super Kings   
2               Rajasthan Royals           bat             Delhi Daredevils   
3                 Mumbai Indians           bat  Royal Challengers Bangalore   
4                Deccan Chargers           bat        Kolkata Knight Riders   
..                           ...           ...                          ...   
811               Mumbai Indians         field  Royal Challengers Bangalore   
812               Delhi Capitals         field               Mumbai Indians   
813          Sunrisers Hyderabad         field          Sunrisers Hyderabad   
814               Delhi Capitals           bat               Delhi Capitals   
815               Delhi Capitals           bat               Mumbai Indians   

      result  result_margin eliminator method      umpire1         umpire2  
0       runs          140.0          N    NaN    Asad Rauf     RE Koertzen  
1       runs           33.0          N    NaN    MR Benson      SL Shastri  
2    wickets            9.0          N    NaN    Aleem Dar  GA Pratapkumar  
3    wickets            5.0          N    NaN     SJ Davis       DJ Harper  
4    wickets            5.0          N    NaN    BF Bowden     K Hariharan  
..       ...            ...        ...    ...          ...             ...  
811      tie            NaN          Y    NaN  Nitin Menon      PR Reiffel  
812     runs           57.0          N    NaN  CB Gaffaney     Nitin Menon  
813  wickets            6.0          N    NaN   PR Reiffel          S Ravi  
814     runs           17.0          N    Na

### ab jahan bhi royal challengers bengaluru hai wahan pe i want RCB then here  convertors come into roles

In [120]:
def rename(name):
    if name=="Royal Challengers Bangalore":
        return "RCB"
    elif name=="Chennai Super Kings":
        return "CSK"
    elif name=="Kolkata Knight Riders":
        return "KKR"
    else:
        return name

In [122]:
rename("Kolkata Knight Riders")

'KKR'

In [131]:
pd.read_csv('IPL Matches 2008-2020.csv',converters={'team1':rename,'team2':rename,'toss_winner':rename,'winner':rename})

id        city        date player_of_match  \
0     335982   Bangalore  2008-04-18     BB McCullum   
1     335983  Chandigarh  2008-04-19      MEK Hussey   
2     335984       Delhi  2008-04-19     MF Maharoof   
3     335985      Mumbai  2008-04-20      MV Boucher   
4     335986     Kolkata  2008-04-20       DJ Hussey   
..       ...         ...         ...             ...   
811  1216547       Dubai  2020-09-28  AB de Villiers   
812  1237177       Dubai  2020-11-05       JJ Bumrah   
813  1237178   Abu Dhabi  2020-11-06   KS Williamson   
814  1237180   Abu Dhabi  2020-11-08      MP Stoinis   
815  1237181       Dubai  2020-11-10        TA Boult   

                                          venue  neutral_venue  \
0                         M Chinnaswamy Stadium              0   
1    Punjab Cricket Association Stadium, Mohali              0   
2                              Feroz Shah Kotla              0   
3                              Wankhede Stadium              0   
4                                  Eden Gardens              0   
..                                          ...            ...   
811         Dubai International Cricket Stadium              0   
812         Dubai International Cricket Stadium              0   
813                        Sheikh Zayed Stadium              0   
814                        Sheikh Zayed Stadium              0   
815         Dubai International Cricket Stadium              0   

                team1                team2          toss_winner toss_decision  \
0                 RCB                  KKR                  RCB         field   
1     Kings XI Punjab                  CSK                  CSK           bat   
2    Delhi Daredevils     Rajasthan Royals     Rajasthan Royals           bat   
3      Mumbai Indians                  RCB       Mumbai Indians           bat   
4                 KKR      Deccan Chargers      Deccan Chargers           bat   
..                ...                  ...                  ...           ...   
811               RCB       Mumbai Indians       Mumbai Indians         field   
812    Mumbai Indians       Delhi Capitals       Delhi Capitals         field   
813               RCB  Sunrisers Hyderabad  Sunrisers Hyderabad         field   
814    Delhi Capitals  Sunrisers Hyderabad       Delhi Capitals           bat   
815    Delhi Capitals       Mumbai Indians       Delhi Capitals           bat   

                  winner   result  result_margin eliminator method  \
0                    KKR     runs          140.0          N    NaN   
1                    CSK     runs           33.0          N    NaN   
2       Delhi Daredevils  wickets            9.0          N    NaN   
3                    RCB  wickets            5.0          N    NaN   
4                    KKR  wickets            5.0          N    NaN   
..                   ...      ...            ...        ...    ...   
811                  RCB      tie            NaN          Y    NaN   
812       Mumbai Indians     runs           57.0          N    NaN   
813  Sunrisers Hyderabad  wickets            6.0          N    NaN   
814       Delhi Capitals     runs           17.0          N    NaN   
815       Mumbai Indians  wickets            5.0          N    NaN   

         umpire1         umpire2  
0      Asad Rauf     RE Koertzen  
1      MR Benson      SL Shastri  
2      Aleem Dar  GA Pratapkumar  
3       SJ Davis       DJ Harper  
4      BF Bowden     K Hariharan  
..           ...             ...  
811  Nitin Menon      PR Reiffel  
812  CB Gaffaney     Nitin Menon  
813   PR Reiffel          S Ravi  
814   PR Reiffel          S Ravi  
815  CB Gaffaney     Nitin Menon  

[816 rows x 17 columns]

# 14.  Handling Missing Values 

# use parameter na_values

In [145]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

###### if i want that all male values should be nan so i can do that by folluowing method

In [147]:
pd.read_csv('aug_train.csv',na_values=['Male'])

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920    NaN   
1            29725   city_40                   0.776    NaN   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767    NaN   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878    NaN   
19154        31398  city_103                   0.920    NaN   
19155        24576  city_103                   0.920    NaN   
19156         5756   city_65                   0.802    NaN   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

# 15. loading huge dataset in chunks

In [151]:
pd.read_csv('aug_train.csv')

enrollee_id      city  city_development_index gender  \
0             8949  city_103                   0.920   Male   
1            29725   city_40                   0.776   Male   
2            11561   city_21                   0.624    NaN   
3            33241  city_115                   0.789    NaN   
4              666  city_162                   0.767   Male   
...            ...       ...                     ...    ...   
19153         7386  city_173                   0.878   Male   
19154        31398  city_103                   0.920   Male   
19155        24576  city_103                   0.920   Male   
19156         5756   city_65                   0.802   Male   
19157        23834   city_67                   0.855    NaN   

           relevent_experience enrolled_university education_level  \
0      Has relevent experience       no_enrollment        Graduate   
1       No relevent experience       no_enrollment        Graduate   
2       No relevent experience    Full time course        Graduate   
3       No relevent experience                 NaN        Graduate   
4      Has relevent experience       no_enrollment         Masters   
...                        ...                 ...             ...   
19153   No relevent experience       no_enrollment        Graduate   
19154  Has relevent experience       no_enrollment        Graduate   
19155  Has relevent experience       no_enrollment        Graduate   
19156  Has relevent experience       no_enrollment     High School   
19157   No relevent experience       no_enrollment  Primary School   

      major_discipline experience company_size    company_type last_new_job  \
0                 STEM        >20          NaN             NaN            1   
1                 STEM         15        50-99         Pvt Ltd           >4   
2                 STEM          5          NaN             NaN        never   
3      Business Degree         <1          NaN         Pvt Ltd        never   
4                 STEM        >20        50-99  Funded Startup            4   
...                ...        ...          ...             ...          ...   
19153       Humanities         14          NaN             NaN            1   
19154             STEM         14          NaN             NaN            4   
19155             STEM        >20        50-99         Pvt Ltd            4   
19156              NaN         <1      500-999         Pvt Ltd            2   
19157              NaN          2          NaN             NaN            1   

       training_hours  target  
0                  36     1.0  
1                  47     0.0  
2                  83     0.0  
3                  52     1.0  
4                   8     0.0  
...               ...     ...  
19153              42     1.0  
19154              52     1.0  
19155              44     0.0  
19156              97     0.0  
19157             127     0.0  

[19158 rows x 14 columns]

### here i have 19158 rows but in future if i have 10000000 rows then i can handle it using or first i will divide the big dataset in small chunks then i will use it accordingly 

In [154]:
dfs=pd.read_csv('aug_train.csv',chunksize=5000)

In [156]:
for chunk in dfs:
    print(chunk.shape)

(5000, 14)
(5000, 14)
(5000, 14)
(4158, 14)


In [166]:
# we can know the size of each chunk by just print(chunk.size)
chunk.size

58212